### ISSUE TITLE
each write request to fuse layer caused one getattr call

### ISSUE THREAD

**Issue by xiaolong302** [billziss-gh/winfsp#326]
## Question

I am moving my linux fuse to use Winfsp. However I noticed that for each say, 4k write, it will trigger a call for "getattr" which degrades performance a lot. I tried options like FileInfoTimeout=1000 or attr_timeout=1, none of them worked. Is there a way to enable attr cache? It seems it should work, but due to this, my write performance is horrible.

**Comment by billziss-gh** [comment]
The WinFsp kernel protocol requires that the file size is correctly reported upon completion of a `Write` request. The WinFsp-FUSE layer (i.e. the FUSE layer of WinFsp) needs to perform an `getattr` or `fgetattr` in order to implement this requirement:

https://github.com/billziss-gh/winfsp/blob/v1.8B3/src/dll/fuse/fuse_intf.c#L1286

Things you may want to try:

- Enable full data caching by using `FileInfoTimeout=-1`. This should greatly diminish the number of `write` (and associated `getattr`) calls that you get.
- Implement `fgetattr` as a faster alternative to `getattr`.

Things that we may want to improve in WinFsp-FUSE:

- All timeouts currently apply to information that lives in the FSD (kernel-mode file system driver). Perhaps it might be worth caching some of this information in user mode as well to eliminate some getattr calls that come from user mode.

**Comment by xiaolong302** [comment]
Thank you for your reply. You are always very helpful.  I tried FileInfoTimeout=-1, it actually significantly slow it down. I am running under parallels on MAC, not sure whether it matters. So what I can conclude is in winfsp, getattr will be called for each write?   FileInfoTimeout/attr_timeout is not able to cache "attr" to avoid making calls to getattr?   Also in linux, we have writeback cache support, it seems from your response, FileInfoTimeout=-1 can achieve similar effect as writeback cache in linux? In linux, writeback cache will batch all 4K IO into 128K buffer and send to user mode.

**Comment by billziss-gh** [comment]
Yes, `FileInfoTimeout=-1` instructs the FSD (File System Driver) to enable the kernel data cache.

This does the following:

- Ensures that kernel mode metadata caches (e.g. file information such as size, dates, etc.) do not expire (unless there is memory pressure or the kernel decides to invalidate them for other reasons).
- Ensures that the kernel uses system memory to cache data and satisfy I/O requests. For example, if the application issues 2 reads for the same file range, the second read will likely be satisfied from cache.
- Enables the read-ahead mechanism, so that the kernel will read large blocks from the file system in anticipation that the application will access those blocks soon.
- Enables the write-behind mechanism, so that application writes are buffered and special system threads will flush cache changes every few seconds (thus greatly reducing the number of `write` requests into your file system).

In general this makes things go considerably faster, so I am somewhat perplexed that this is not the case for you.

> So what I can conclude is in winfsp, getattr will be called for each write?

Unfortunately in the current WinFsp-FUSE implementation this is indeed the case for the reasons explained in my earlier message. I definitely think it is worth exploring user mode caching of `attr` information to avoid this problem.

**Comment by xiaolong302** [comment]
OK, I will probably need to optimize the getattr implementation to make it more performant. I tried to run with `FileInfoTimeout=-1,` in order to rule out other possibilities, I tried passthrough-fuse3 in your tst in one of windows EC2 instance. it turns out that without setting FileInfoTimeout=-1, I can achieve about 100 MB/s using fio with 4K IO size. With `FileInfoTimeout=-1,` the throughput degraded to 50 MB. It really looks like setting   `FileInfoTimeout=-1` has negative impact. I am using latest winfsp and passthrough-fuse3 in your repo. I guess you can probably do a quick run to see yourself if it's convenient for you. fio cmd is `fio --rw=write --bs=4k --iodepth=16 --numjobs=1 --name=test --size=1G --fallocate=0`.  fio can be downloaded and installed here: https://bsdio.com/fio/

Lastly, even `FileInfoTimeout=-1` works as expected, is there a way to manually flush cache to underling file system. We implement in a stackable file system way, so we want to flush all bytes to underling file system and  then upload underling files to cloud. In linux, I am using `fuse_invalidate_path` method

**Comment by billziss-gh** [comment]
> Lastly, even FileInfoTimeout=-1 works as expected, is there a way to manually flush cache to underling file system. We implement in a stackable file system way, so we want to flush all bytes to underling file system and then upload underling files to cloud. In linux, I am using fuse_invalidate_path method

This capability was added recently and it is available via a mechanism which can issue both cache invalidations and file change notifications. I added commit 1b7a78edffa6890d811b5e7d00778a2cc02b47a0 which makes the FUSE2 `fuse_invalidate` available. This work is in private branch [pvt-notify](https://github.com/billziss-gh/winfsp/tree/pvt-notify) and has not been merged into master yet.
